### PR TITLE
IACD Test Fix

### PR DESCRIPTION
**Pull request by DStavilaNI**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-instrument-addon-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Refactoring for the committed tests. Fixes the issue with the intermittent failures on the ATS system.

### Why should this Pull Request be merged?

- Fixes the issue with the intermittent failures on the ATS system.

### What testing has been done?

- Testing has been done on a VM with the 2020 software stack installed + RT Pharlap Target

### CODE REVIEW

**Review by phutanu — CHANGES_REQUESTED** [review]
The two tests from below were validating that a Periodic Message was triggered after the set time has elapsed. The refactoring does not validate this anymore. As you can notice on both Block Diagrams of these test VIs we trigger "N" messages, but we read "N+1" in the "Expected Messages" array, as we are expecting to have one Periodic Message triggered. Also the content of the Periodic Message is not checked anymore.
From my point of view, this needs to be addressed, as we may have a test gap using this proposed implementation.

- test_Periodic and Multiple Static Single Channel Command-Response Messages using Trigger with Protection.vi
- test_Periodic and Multiple Static Single Channel Command-Response messages using Trigger.vi

**Comment by XenderMd** [comment]
> 
> 
> The two tests from below were validating that a Periodic Message was triggered after the set time has elapsed. The refactoring does not validate this anymore. As you can notice on both Block Diagrams of these test VIs we trigger "N" messages, but we read "N+1" in the "Expected Messages" array, as we are expecting to have one Periodic Message triggered. Also the content of the Periodic Message is not checked anymore.
> From my point of view, we need to address this, as we may add a test gap using the proposed implementation.
> 
>     * test_Periodic and Multiple Static Single Channel Command-Response Messages using Trigger with Protection.vi
> 
>     * test_Periodic and Multiple Static Single Channel Command-Response messages using Trigger.vi

It was not obvious in any way, from the original test, that a periodical message is actually expected and checked (even though, indeed, the System Definition is actually configured with one). The way the content was being checked is completely fragile and it was actually leading to the test failures reported before. The idea would be to separate the tests for the triggered messages from the periodical one - which means creating a separate test only for the periodical message.

**Comment by phutanu** [comment]
> > The two tests from below were validating that a Periodic Message was triggered after the set time has elapsed. The refactoring does not validate this anymore. As you can notice on both Block Diagrams of these test VIs we trigger "N" messages, but we read "N+1" in the "Expected Messages" array, as we are expecting to have one Periodic Message triggered. Also the content of the Periodic Message is not checked anymore.
> > From my point of view, we need to address this, as we may add a test gap using the proposed implementation.
> > ```
> > * test_Periodic and Multiple Static Single Channel Command-Response Messages using Trigger with Protection.vi
> > 
> > * test_Periodic and Multiple Static Single Channel Command-Response messages using Trigger.vi
> > ```
> 
> It was not obvious in any way, from the original test, that a periodical message is actually expected and checked (even though, indeed, the System Definition is actually configured with one). The way the content was being checked is completely fragile and it was actually leading to the test failures reported before. The idea would be to separate the tests for the triggered messages from the periodical one - which means creating a separate test only for the periodical message.

We already have a test for the Periodic Channel and Periodic Message. So that is covered.
As this two test names imply, they check that a Periodic Message is sent even if Multiple Static Single Channels Messages are Triggered without affecting the Periodicity.

We are here to improve the test logic, but also to keep the current code coverage as before, or improving it if we can.
I would like to not discard tests that test this scenario (sending multiple triggered messages with periodic messages in between them). I do not mind if we do this adding new tests, or try to improve the ones in this PR.

**Comment by XenderMd** [comment]
> 
> 
> > > The two tests from below were validating that a Periodic Message was triggered after the set time has elapsed. The refactoring does not validate this anymore. As you can notice on both Block Diagrams of these test VIs we trigger "N" messages, but we read "N+1" in the "Expected Messages" array, as we are expecting to have one Periodic Message triggered. Also the content of the Periodic Message is not checked anymore.
> > > From my point of view, we need to address this, as we may add a test gap using the proposed implementation.
> > > ```
> > > * test_Periodic and Multiple Static Single Channel Command-Response Messages using Trigger with Protection.vi
> > > 
> > > * test_Periodic and Multiple Static Single Channel Command-Response messages using Trigger.vi
> > > ```
> > 
> > 
> > It was not obvious in any way, from the original test, that a periodical message is actually expected and checked (even though, indeed, the System Definition is actually configured with one). The way the content was being checked is completely fragile and it was actually leading to the test failures reported before. The idea would be to separate the tests for the triggered messages from the periodical one - which means creating a separate test only for the periodical message.
> 
> We already have a test for the Periodic Channel and Periodic Message. So that is covered.
> As this two test names imply, they check that a Periodic Message is sent even if Multiple Static Single Channels Messages are Triggered without affecting the Periodicity.
> 
> We are here to improve the test logic, but also to keep the current code coverage as before, or improving it if we can.
> I would like to not discard tests that test this scenario (sending multiple triggered messages with periodic messages in between them). I do not mind if we do this adding new tests, or try to improve the ones in this PR.

Ok ... than I think that this should have been explained in a comment, in the VI itself. The title of the VI was not very suggestive as to how the test was actually operating.

**Comment by phutanu** [comment]
> Ok ... than I think that this should have been explained in a comment, in the VI itself. The title of the VI was not very suggestive as to how the test was actually operating. The test, as it was, did not validate in anyway the periodicity of the message since it was checking if the periodic message was received only once ... and to be fair, the periodicity in this particular case could be checked only approximately. What the original tests were actually checking, was simply if the message was sent once, not in any way if it is periodic.

We should operate changes to code only when we fully understand the implications and ask if otherwise.
As you may know, we currently do not have an official guideline, so is at the developers willingness on how he decides to document the code if not requested explicitly in the task or US. You are free to add any additional info you think may help in the future.

As I explained before, this tests were checking if the periodicity of the message was affected by triggering other messages. The IA CD is made to run in tens of Hz frequency ranges, and we do not expect and also do not guarantee exact timings, as there are a lot of factors that can affect the timing like: the LabVIEW Version, Target Type, VeriStand System latency, VeriStand System size, the no of Instrument Addon Messages we send, the Protocol we use, etc. 
So for this setup, the decision was to have 4 messages sent before getting the Periodic one. And that was what we were checking. And it seems was not the best approach for this particular tests.

**Comment by XenderMd** [comment]
> 
> 
> > Ok ... than I think that this should have been explained in a comment, in the VI itself. The title of the VI was not very suggestive as to how the test was actually operating. The test, as it was, did not validate in anyway the periodicity of the message since it was checking if the periodic message was received only once ... and to be fair, the periodicity in this particular case could be checked only approximately. What the original tests were actually checking, was simply if the message was sent once, not in any way if it is periodic.
> 
> We should operate changes to code only when we fully understand the implications and ask if otherwise.
> As you may know, we currently do not have an official guideline, so is at the developers willingness on how he decides to document the code if not requested explicitly in the task or US. You are free to add any additional info you think may help in the future.
> 
> As I explained before, this tests were checking if the periodicity of the message was affected by triggering other messages. The IA CD is made to run in tens of Hz frequency ranges, and we do not expect and also do not guarantee exact timings, as there are a lot of factors that can affect the timing like: the LabVIEW Version, Target Type, VeriStand System latency, VeriStand System size, the no of Instrument Addon Messages we send, the Protocol we use, etc.
> So for this setup, the decision was to have 4 messages sent before getting the Periodic one. And that was what we were checking. And it seems was not the best approach for this particular tests.

To your point - I made the changes in the code based on what I saw was done in the original code. The purpose of the PR is exactly to validate this kind of changes and solve questions if they appear, so I think there should be no issue with my understanding. Since the test did not validate in any way the periodicity ("periodicity" meaning in this case that the message is periodic) of the periodic message, again, since it was received only once during the original test, and combined with the lack of documentation, my conclusion was that it was simply not required or that it was a "by-product" of the test.

**Comment by phutanu** [comment]
This test checks that the periodic message is sent, but it does not check if Triggering other messages affect the timing of the Periodic Message. We check the similar scenario in "test_Enable Periodic Message Channel.vi". So this test is redundant if it does not check the additional Triggering of other messages and if this affects the Periodic one.

image

**Comment by XenderMd** [comment]
> 
> 
> This test checks that the periodic message is sent, but it does not check if Triggering other messages affect the timing of the Periodic Message. We check the similar scenario in "test_Enable Periodic Message Channel.vi". So this test is redundant if it does not check the additional Triggering of other messages and if this affects the Periodic one.
> 
> image

I was in the process of modifying this test based on your last comments.

**Comment by phutanu** [comment]
> > This test checks that the periodic message is sent, but it does not check if Triggering other messages affect the timing of the Periodic Message. We check the similar scenario in "test_Enable Periodic Message Channel.vi". So this test is redundant if it does not check the additional Triggering of other messages and if this affects the Periodic one.
> > image
> 
> I was in the process of modifying this test based on your last comments.

Great! Thanks! Will wait for the diffs.

**Comment by phutanu** [comment]
test_Multiple Periodic Command Message.vi

The 16 second interval is really close to the border, as we do not expect to have exact timing. So I consider we should send like 20 messages (aprox 18-20 seconds IA runtime) to make sure we get also the second message in a specified range.

- 1st periodic message ~8+ sec
- 2nd periodic message ~16+ sec
- 3rd periodic message ~ 24+ sec

So if we stick to a range around 19-21 seconds runtime, we can make sure we have the second periodic message, but the chances to get a third one, are pretty slim. The current setup having 16 seconds runtime may make this test intermittent, as we may not get the second periodic message.

Please add also a check using the "Check Message Occurrences.vi" to confirm that we get also the expected number of triggered messages having the below content.
image

I would also preferred to have a way of also checking the order in which we get the messages, but for this specific test, I do not think is required.

image

**Comment by DStavilaNI** [comment]
> 
> 
> test_Multiple Periodic Command Message.vi
> 
> The 16 second interval is really close to the border, as we do not expect to have exact timing. So I suggest we should send 20 messages (aprox 18-20 seconds IA runtime) to make sure we get also the second message in the specified range.
> 
>     * 1st periodic message ~8+ sec
> 
>     * 2nd periodic message ~16+ sec
> 
>     * 3rd periodic message ~ 24+ sec
> 
> 
> So if we stick to a range around 19-21 seconds runtime, we can make sure we have the second periodic message, but know we should not get a third one. The current setup is having a 16 seconds aprox runtime and this may make the test intermittent, as we may not get the second periodic message.
> 
> Please add also a check using the "Check Message Occurrences.vi" to confirm that we get the expected number of triggered messages having the below content.
> image
> 
> I would also preferred to have a way of checking the order in which we get the messages, but for this specific test, I do not think it makes any sense.
> 
> image

I think that the time is more than enough - each iteration of the "For Loop" takes much more than a second due to the execution time of the trigger-confirm function. Moreover, there are delays between when the deployment is finished and when the triggering execution starts. During the development and test phase for this VI, always, the first periodic message would arrive before the sixth triggered command - this gives an excess of 2 seconds of margin. I agree however with adding 2 more iterations to be on the safe side. 

I will not add the extra testing for the triggered messages - this is not the purpose of this test, additionally it would be redundant since we have dedicated tests for this behavior, for these specific channels.

**Comment by phutanu** [comment]
>  I agree however with adding 2 more iterations to be on the safe side.
👍 


> I will not add the extra testing for the triggered messages - this is not the purpose of this test, additionally it would be redundant since we have dedicated tests for this behavior, for these specific channels.
I do not think the additional check is redundant, as the purpose of this test is to make sure that this messages work together independent on what trigger is used (trigger channel, periodic). So that's why I think we should check we have the expected number and content for both received message types. Only then we can guarantee they work as expected.

The additional check implementation is trivial, and should not require a lot of extra work, as you already have the VIs needed to do this.

Thanks!

**Comment by phutanu** [comment]
The highlighted comment is not valid anymore, as we now check only a specific message content and no of occurrences.

image

**Comment by DStavilaNI** [comment]
> 
> 
> > I agree however with adding 2 more iterations to be on the safe side.
> 
> 👍
> 
> > I will not add the extra testing for the triggered messages - this is not the purpose of this test, additionally it would be redundant since we have dedicated tests for this behavior, for these specific channels.
> 
> I do not think the additional check is redundant, as the purpose of this test is to make sure that this messages work together independent on what trigger is used (trigger channel, periodic). So that's why I think we should check we have the expected number and content for both received message types. Only then we can guarantee they work as expected.
> 
> The additional check implementation is trivial, and should not require a lot of extra work, as you already have the VIs needed to do this.
> 
> Thanks!

We have already a test that validates that the triggered command messages are actually being sent. Would it be possible for you to explain why this extra measure of certainty is required for this test ?

**Comment by phutanu** [comment]
> We have already a test that validates that the triggered command messages are actually being sent, and implicitly that this functionality works as expected - verified. Why is it not enough ? Would it be possible for you to explain why this extra measure of certainty is required for this test ?

As I explained in previous messages is that this test needs to validate that triggering a message does not affect a periodic one. But that goes both ways. So, if a periodic message was triggered, we want to make sure the triggered one still works fine.

Basically, we want to make sure a triggered and a periodic message work ok together. That's why we need to check both.

**Comment by DStavilaNI** [comment]
> 
> 
> > We have already a test that validates that the triggered command messages are actually being sent, and implicitly that this functionality works as expected - verified. Why is it not enough ? Would it be possible for you to explain why this extra measure of certainty is required for this test ?
> 
> As I explained in a previous message is that this test needs to validate that triggering a message does not affect a periodic one. But that goes both ways. So, if a periodic message was triggered, we want to make sure the triggered one still works fine.
> 
> Basically, we want to make sure a triggered and a periodic message work ok together. That's why we need to check both.

The periodic message is enabled default in the system definition - this means that periodic messages are being sent, once the deployment is complete, for all of the tests. This also means that periodic messages are being sent anyways during the execution of the tests for the triggered command messages - so I would imagine that this aspect is also covered by the other dedicated tests. Or you think that this is not the case ?

**Comment by phutanu** [comment]
> > > We have already a test that validates that the triggered command messages are actually being sent, and implicitly that this functionality works as expected - verified. Why is it not enough ? Would it be possible for you to explain why this extra measure of certainty is required for this test ?
> > 
> > 
> > As I explained in a previous message is that this test needs to validate that triggering a message does not affect a periodic one. But that goes both ways. So, if a periodic message was triggered, we want to make sure the triggered one still works fine.
> > Basically, we want to make sure a triggered and a periodic message work ok together. That's why we need to check both.
> 
> Thank you for the explanation. I would like to point out that the periodic message is enabled by default in the system definition - this means that periodic messages are being sent, once the deployment is complete, for all of the tests. This also means that periodic messages are being sent anyways during the execution of the tests for the triggered command messages - so I would imagine that this aspect is also covered by the other dedicated tests. Or you think that this is not the case ?

You are right, the fact that the periodic message is generated is covered as a side topic in other tests. But there is no dedicated test that actually checks if the periodicity is correct, or if the number of expected periodic messages are generated in a defined timeframe.

**Comment by DStavilaNI** [comment]
> 
> 
> > > > We have already a test that validates that the triggered command messages are actually being sent, and implicitly that this functionality works as expected - verified. Why is it not enough ? Would it be possible for you to explain why this extra measure of certainty is required for this test ?
> > > 
> > > 
> > > As I explained in a previous message is that this test needs to validate that triggering a message does not affect a periodic one. But that goes both ways. So, if a periodic message was triggered, we want to make sure the triggered one still works fine.
> > > Basically, we want to make sure a triggered and a periodic message work ok together. That's why we need to check both.
> > 
> > 
> > Thank you for the explanation. I would like to point out that the periodic message is enabled by default in the system definition - this means that periodic messages are being sent, once the deployment is complete, for all of the tests. This also means that periodic messages are being sent anyways during the execution of the tests for the triggered command messages - so I would imagine that this aspect is also covered by the other dedicated tests. Or you think that this is not the case ?
> 
> You are right, the fact that the periodic message is generated is covered as a side topic in other tests. But there is no dedicated test that actually checks if the periodicity is correct, or if the number of expected periodic messages are generated in a defined timeframe.

I think that it would benefit us to clarify things a little bit. This is a dedicated test (test_Multiple Periodic Command Message.vi) that verifies that a periodic message is being generated, checks it's period (approximately of course) and that it works even if we are triggering additional command messages. I am not sure what do you mean exactly when you are saying " testing the periodicity", but assuming that you mean testing that the target command message is generated periodically - than this test does exactly that. 

Additionally, since we have tests that trigger multiple command messages, while the periodic command message is enabled (by default in the system definition), it means that we are testing also that a triggered command message can execute while a periodical command message is also running.  Would it be possible for you to point out what aspect exactly you think is missing ? Thank you.

**Comment by phutanu** [comment]
> I think that it would benefit us to clarify things a little bit. This is a dedicated test (test_Multiple Periodic Command Message.vi) that verifies that a periodic message is being generated, checks it's period (approximately of course) and that it works even if we are triggering additional command messages. I am not sure what do you mean exactly when you are saying " testing the periodicity", but assuming that you mean testing that the target command message is generated periodically - than this test does exactly that.

Yes, I agree with you on the above summary regarding the current implementation of the test. And I agree that the current implementation of the test is validating the periodicity of the message in the expected runtime.

Testing the periodicity = testing the tendency of the periodic message to recur at intervals as per the term definition.

> Additionally, since we have tests that trigger multiple command messages, while the periodic command message is enabled (by default in the system definition), it means that we are testing also that a triggered command message can execute while a periodical command message is also running. Would it be possible for you to point out what aspect exactly you think is missing ? Thank you.

In previous tests we just check if we get a periodic message, but just as a side effect of the periodic message being enabled by default.

So, long story short, if this test will have also a check that the number and content of the triggered message is as we expect, only then we can say for sure that the Messages work independently with no influence from the trigger type.

**Comment by DStavilaNI** [comment]
> 
> 
> > I think that it would benefit us to clarify things a little bit. This is a dedicated test (test_Multiple Periodic Command Message.vi) that verifies that a periodic message is being generated, checks it's period (approximately of course) and that it works even if we are triggering additional command messages. I am not sure what do you mean exactly when you are saying " testing the periodicity", but assuming that you mean testing that the target command message is generated periodically - than this test does exactly that.
> 
> Yes, I agree with you on the above summary regarding the current implementation of the test. And I agree that the current implementation of the test is validating the periodicity of the message in the expected runtime.
> 
> Testing the periodicity = testing the tendency of the periodic message to recur at intervals as per the term definition.
> 
> > Additionally, since we have tests that trigger multiple command messages, while the periodic command message is enabled (by default in the system definition), it means that we are testing also that a triggered command message can execute while a periodical command message is also running. Would it be possible for you to point out what aspect exactly you think is missing ? Thank you.
> 
> In previous tests we just check if we get a periodic message, but just as a side effect of the periodic message being enabled by default.
> 
> So, long story short, if this test will have also a check that the number and content of the triggered message that is triggered by the Trigger Channel is as we expect, only then we can say for sure that the Messages work independently with no influence from the trigger type used.
> Thanks!

Thank you for the reply. I don't see how this last comment, or paragraph is an argument - it seems to be reiterating the request for change. At the same time, I also believe that it does not make too much sense to continue this discussion. If you deem it necessary, I will apply this modification.

**Comment by phutanu** [comment]
> > > I think that it would benefit us to clarify things a little bit. This is a dedicated test (test_Multiple Periodic Command Message.vi) that verifies that a periodic message is being generated, checks it's period (approximately of course) and that it works even if we are triggering additional command messages. I am not sure what do you mean exactly when you are saying " testing the periodicity", but assuming that you mean testing that the target command message is generated periodically - than this test does exactly that.
> > 
> > 
> > Yes, I agree with you on the above summary regarding the current implementation of the test. And I agree that the current implementation of the test is validating the periodicity of the message in the expected runtime.
> > Testing the periodicity = testing the tendency of the periodic message to recur at intervals as per the term definition.
> > > Additionally, since we have tests that trigger multiple command messages, while the periodic command message is enabled (by default in the system definition), it means that we are testing also that a triggered command message can execute while a periodical command message is also running. Would it be possible for you to point out what aspect exactly you think is missing ? Thank you.
> > 
> > 
> > In previous tests we just check if we get a periodic message, but just as a side effect of the periodic message being enabled by default.
> > So, long story short, if this test will have also a check that the number and content of the triggered message that is triggered by the Trigger Channel is as we expect, only then we can say for sure that the Messages work independently with no influence from the trigger type used.
> > Thanks!
> 
> Thank you for the reply. I don't see how this last comment, or paragraph is an argument - it seems to be reiterating the request for change. At the same time, I also believe that it does not make too much sense to continue this discussion. If you deem it necessary, I will apply this modification.



> > > I think that it would benefit us to clarify things a little bit. This is a dedicated test (test_Multiple Periodic Command Message.vi) that verifies that a periodic message is being generated, checks it's period (approximately of course) and that it works even if we are triggering additional command messages. I am not sure what do you mean exactly when you are saying " testing the periodicity", but assuming that you mean testing that the target command message is generated periodically - than this test does exactly that.
> > 
> > 
> > Yes, I agree with you on the above summary regarding the current implementation of the test. And I agree that the current implementation of the test is validating the periodicity of the message in the expected runtime.
> > Testing the periodicity = testing the tendency of the periodic message to recur at intervals as per the term definition.
> > > Additionally, since we have tests that trigger multiple command messages, while the periodic command message is enabled (by default in the system definition), it means that we are testing also that a triggered command message can execute while a periodical command message is also running. Would it be possible for you to point out what aspect exactly you think is missing ? Thank you.
> > 
> > 
> > In previous tests we just check if we get a periodic message, but just as a side effect of the periodic message being enabled by default.
> > So, long story short, if this test will have also a check that the number and content of the triggered message that is triggered by the Trigger Channel is as we expect, only then we can say for sure that the Messages work independently with no influence from the trigger type used.
> > Thanks!
> 
> Thank you for the reply. I don't see how this last comment, or paragraph is an argument - it seems to be reiterating the request for change. At the same time, I also believe that it does not make too much sense to continue this discussion. If you deem it necessary, I will apply this modification.

Yes, I consider it necessary for validating both message types work together as expected.
Thanks!

**Review by phutanu — APPROVED** [review]

**Review by bogdanp-ni — APPROVED** [review]
